### DISCUSSION TITLE
How QR Data only encrypted  in digits

### DISCUSSION THREAD

**Discussion by lalwanikamaldev** [tanmoysrt/pyaadhaar#13] [General]
Hey , i am very curious about  how the QR encrypted data only shows the Digits only when data get encrypted.
which **algo they used so that they get only digits in data** .

I was reading this manual 
https://103.57.226.101/images/resource/User_manulal_QR_Code_15032019.pdf

I was seeing the sample data it shows always a digits 
<img width="194" alt="image" src="https://github.com/tanmoysrt/pyaadhaar/assets/65231431/16f25fd1-e32f-491c-969c-3b62b927db4a">

https://www.appsloveworld.com/coding/python3x/21/how-to-read-decode-secure-qr-code-on-indian-aadhaar-card-image 

In this article also simple data is digits only .

Even when i use paytm qr app to scan the adhar barcode ,it shows me only digits .

My questions is how the json data get converted to only digits and which algorithm they use to do that .

**Comment by tanmoysrt** [comment]
Hi @lalwanikamaldev 

In the case of Paytm or others, they generate randomize data maximum time.

In the case of Aadhaar, it needs to be highly compressed to fit in a small size, so here is the technique they use
- Take main data [maybe JSON, XML, YML, TOML]
- Convert to string
- Convert the string data to bytes array
- Compress that bytes array by any algo [zlib is a good library for this]
- Encode compressed bytes array by base10 encoding so that we can write to QR
- Base10 is decimal stuff [0~9], that's the reason you see only digits in this type of QR

**Reply by lalwanikamaldev** [reply]
I have image which 1200 byte and after converting it to base 64 string it becomes 1700 bytes and after that i am adding data to it and it become 2100 bytes after that if i compress it , it become 1700 Bytes . then i encrypted it and it become 2200 bytes .

once i convert 2200 bytes to 4000 bytes .

do you have any idea how we can compress it more .
